File: 4_patrones_array_strings/13-longest-substring.js
function lengthOfLongestSubstring(letras) {
   let maxSize = 0;
   let p1 = 0;
   let map = new Map();
   for (let p2 = 0; p2 < letras.length; p2++) {
      const actualindex = map.get(letras[p2])
      if (actualindex == undefined || actualindex < p1) {
         map.set(letras[p2], p2)
         const newMaxSize = p2 - p1 + 1;
         if (newMaxSize > maxSize) {
            maxSize = newMaxSize;
         }
      } else {
         p1 = actualindex + 1
         map.set(letras[p2], p2)
      }
   }
   return maxSize;
}

console.log(lengthOfLongestSubstring("abcabcbb"))
console.log(lengthOfLongestSubstring("jdkafnlcdsalkxcmpoiuytfccv"))

File: 4_patrones_array_strings/13-longest-substring.py
def lengthOfLongestSubstring(s: str) -> int:
  inicio = 0 #apuntador uno
  caracteresAposicion = {} #cambia el reinicio del apuntador uno
  mayorLongitud = 0
  
  for fin in range(len(s)):  # apuntador dos: fin
    if s[fin] in caracteresAposicion and inicio <= caracteresAposicion[s[fin]]: #es un duplicado y inicio <= donde está la posición final
      #inicio = s[fin] + 1
      inicio = caracteresAposicion[s[fin]] + 1
    caracteresAposicion[s[fin]] = fin
    mayorLongitud = max(mayorLongitud, fin - inicio + 1)
  return mayorLongitud


print(lengthOfLongestSubstring('abcabcbb'))
print(lengthOfLongestSubstring('jdkafnlcdsalkxcmpoiuytfccv'))


